File: C++/Kata.LoanPrediction.C++/Kata.LoanPrediction.C++/LoanCalculator.cpp
//
//  LoanCalculator.cpp
//  Kata.LoanPrediction.C++
//
//  Created by a.j.dowding on 3/09/2015.
//  Copyright (c) 2015 a.j.dowding. All rights reserved.
//

#include "LoanCalculator.h"
#include "Global.h"

/*
 ***************************************
 ** Default constructor.
 ***************************************
 */
LoanCalculator::LoanCalculator(LoanContext context)
{
	this->context = context;
}

/*
 ***************************************
 ** Calculates the transactions on the
 ** loan from now to the time it will be
 ** paid off.
 ***************************************
 */
LoanCalculationOutput LoanCalculator::calculateLoan()
{
	LoanCalculationOutput output;
	bool firstIteration = true;
	double balance = context.getBalance();
	double monthlyInterest = 0.0;
	dateTime currentDate = context.getStartDate();
	// we will calculate interest from the first of the month
	dateTime calcInterestStartDate = currentDate;
	calcInterestStartDate.tm_mday = 1;
	output.setInterestStartDate(calcInterestStartDate);
	// loop through dates, caculating interest daily etc
	while (true)
	{
		// if min repay day
		if (currentDate.tm_mday == context.getMinRepaymentDay() && balance > 0.0)
		{
			processMinRepayment(balance, context.getMinRepaymentAmount(), currentDate, monthlyInterest, output);
		}
		// if extra repay day
		if (currentDate.tm_mday == context.getExtraRepaymentDay() && context.getExtraRepaymentAmount() > 0.0 && balance > 0.0)
		{
			processExtraRepayment(balance, context.getExtraRepaymentAmount(), currentDate, output);
		}

		// if balance is zero or less - we are done!
		if (balance <= 0.0)
        {
            output.setLoanEndsDate(currentDate);
            break;
        }
		// calculate the daily interest
		monthlyInterest += calculateDailyInterest(balance, context.getInterestRate());
		// if this is the first iteration and we did not start on the first of the month
		// let's calculate the interest for the days back to the first
		if (firstIteration && context.getStartDate().tm_mday != 1)
		{
			int daysToMonthStart = currentDate.tm_mday - 1;
			monthlyInterest += (monthlyInterest * daysToMonthStart);
			firstIteration = false;
		}
		// move date forward
		dateTime nextDate = currentDate;
		nextDate.tm_mday = nextDate.tm_mday + 1;
		mktime(&nextDate);
		// if end of month
		if (currentDate.tm_mon != nextDate.tm_mon)
		{
			processEndOfMonth(currentDate, monthlyInterest, balance, output);
		}
		currentDate = nextDate;
	}
	setTargetEndDateAccuracy(output.getLoanEndsDate(), context.getTargetEndDate(), output);
	return(output);
}

/*
 ***************************************
 ** Processes the end of month, adds
 ** interest to the balance.
 ***************************************
 */
void LoanCalculator::processEndOfMonth(dateTime currentDate, double &monthlyInterest, double &balance, LoanCalculationOutput &output)
{
	// add interest charged transaction
	balance += monthlyInterest;
	// calculate the total interest paid
	output.setTotalInterestPaid(output.getTotalInterestPaid() + monthlyInterest);
	output.addTransaction(createTransaction(TX_TYPE_INTEREST_CHARGED, currentDate, 0.0, monthlyInterest, balance));
	// reset for new month
	monthlyInterest = 0.0;
}

/*
 ***************************************
 ** Processes the extra repayment for the
 ** current date if applicable.
 ***************************************
 */
void LoanCalculator::processExtraRepayment(double &balance, double extraRepaymentAmount, dateTime currentDate, LoanCalculationOutput &output)
{
	// add extra repay transaction (if amount is more than 0)
	balance -= extraRepaymentAmount;
	output.addTransaction(createTransaction(TX_TYPE_EXTRA_REPAYMENT, currentDate, extraRepaymentAmount, 0.0, balance));
}

/*
 ***************************************
 ** Processes the min repayment for the
 ** current date if applicable.
 ***************************************
 */
void LoanCalculator::processMinRepayment(double &balance, double minRepaymentAmount, dateTime &currentDate, double monthlyInterest, LoanCalculationOutput &output)
{
	// add min repay transaction
	if ((balance + monthlyInterest) <= minRepaymentAmount)
	{
		double finalRepayment = 0.0;
		balance += monthlyInterest;
		finalRepayment = balance;
		balance -= finalRepayment;
		output.addTransaction(createTransaction(TX_FINAL_REPAYMENT, currentDate, finalRepayment, monthlyInterest, balance));
	}
	else
	{
		balance -= context.getMinRepaymentAmount();
		output.addTransaction(createTransaction(TX_TYPE_MIN_REPAYMENT, currentDate, minRepaymentAmount, 0.0, balance));
	}
}

/*
 ***************************************
 ** Calculates the interest for a day.
 ***************************************
 */
double LoanCalculator::calculateDailyInterest(double balance, double interestRate)
{
	double result = 0.0;
	result = (balance * (interestRate / 100)) / NUM_DAYS_IN_YEAR;
	return(result);
}

/*
 ***************************************
 ** Creates a transaction using the
 ** provided values.
 ***************************************
 */
LoanTransaction LoanCalculator::createTransaction(string type, dateTime date, double repayment, double charge, double balance)
{
	LoanTransaction transaction;
	transaction.setType(type);
	transaction.setDate(date);
	transaction.setCredit(repayment);
	transaction.setDebit(charge);
	transaction.setBalance(balance);
	return(transaction);
}

/*
 ***************************************
 ** Sets the accuracy of the target end
 ** end date with respect to the actual
 ** end date.
 ***************************************
 */
void LoanCalculator::setTargetEndDateAccuracy(dateTime actualEndDate, dateTime targetEndDate, LoanCalculationOutput &output)
{
	int daysMissed = 0;
	bool searchForwards = false;
	output.setTargetEndDateHit(false);
	dateTime *search = new dateTime(actualEndDate);
	int compareResult = compareDates(actualEndDate, targetEndDate);
	if (compareResult == 0)
	{
		output.setTargetEndDateHit(true);
		return;
	}
	if (compareResult == 1)
	{
		searchForwards = false;
	}
	else if (compareResult == -1)
	{
		searchForwards = true;
	}
	while (compareDates(*search, targetEndDate) != 0)
	{
		if (searchForwards)
		{
			search->tm_mday = search->tm_mday + 1;
			daysMissed--;
		}
		else
		{
			search->tm_mday = search->tm_mday - 1;
			daysMissed++;
		}
		mktime(search);
	}
	output.setTargetEndDateMissedInDays(daysMissed);
	delete(search);
}

File: C#/Kata.LoanPrediction.C#/Kata.LoanPrediction.CSharp.Common/Calculator/LoanCalculator.cs
using Kata.LoanPrediction.CSharp.Common.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Kata.LoanPrediction.CSharp.Common.Calculator
{
    /// <summary>
    /// Calculates the loan from start date
    /// to a point in time it is paid off.
    /// </summary>
    public class LoanCalculator
    {
        private LoanContext context = null;

        /// <summary>
        /// Initializes a new instance of the <see cref="LoanCalculator"/> class.
        /// </summary>
        /// <param name="context">The context.</param>
        public LoanCalculator(LoanContext context)
        {
            this.context = context;
        }

        /// <summary>
        /// Calculates the loan using the context.
        /// </summary>
        /// <returns></returns>
        public LoanCalculationOutput Calculate()
        {
            LoanCalculationOutput result = new LoanCalculationOutput();
            DateTime nextDate;
            bool firstIteration = true;
            float balance = context.Balance;
            float monthlyInterest = 0.0f;
            DateTime currentDate = context.StartDate;
            // we will calculate interest from the first of the month
            DateTime calcInterestStartDate = new DateTime(currentDate.Year, currentDate.Month, 1);
            result.InterestStartDate = calcInterestStartDate;
            // loop through dates, caculating interest daily etc
            while (true)
            {
                // if min repay day
                if (currentDate.Day == context.MinRepaymentAmount && balance > 0.0f)
                {
                    ProcessMinRepayment(ref balance, context.MinRepaymentAmount, currentDate, monthlyInterest, result);
                }
                // if extra repay day
                if (currentDate.Day == context.ExtraRepaymentDay && context.ExtraRepaymentAmount > 0.0f && balance > 0.0f)
                {
                    ProcessExtraRepayment(ref balance, context.ExtraRepaymentAmount, currentDate, result);
                }

                // if balance is zero or less - we are done!
                if (balance <= 0.0f) break;
                // calculate the daily interest
                monthlyInterest += CalculateDailyInterest(balance, context.InterestRate);
                // if this is the first iteration and we did not start on the first of the month
                // let's calculate the interest for the days back to the first
                if (firstIteration && context.StartDate.Day != 1)
                {
                    int daysToMonthStart = currentDate.Day - 1;
                    monthlyInterest += (monthlyInterest * daysToMonthStart);
                    firstIteration = false;
                }
                // move date forward
                nextDate = currentDate.AddDays(1);
                // if end of month
                if (currentDate.Month != nextDate.Month)
                {
                    ProcessEndOfMonth(currentDate, monthlyInterest, balance, result);
                    // reset interest to zero
                    monthlyInterest = 0.0f;
                }
                currentDate = nextDate;
            }
            SetTargetEndDateAccuracy(result.LoanEndsDate, context.TargetEndDate, result);
            return (result);
        }

        /// <summary>
        /// Processes the end of month.
        /// </summary>
        /// <param name="currentDate">The current date.</param>
        /// <param name="monthlyInterest">The monthly interest.</param>
        /// <param name="balance">The balance.</param>
        /// <param name="result">The result.</param>
        private void ProcessEndOfMonth(DateTime currentDate, float monthlyInterest, float balance, LoanCalculationOutput result)
        {
            // add interest charged transaction
            balance += monthlyInterest;
            // calculate the total interest paid
            result.TotalInterestPaid += monthlyInterest;
            LoanTransaction transaction = new LoanTransaction(currentDate, TransactionType.InterestCharged, 0.0f, monthlyInterest, balance);
            result.Transactions.Add(transaction);
        }

        /// <summary>
        /// Processes the extra repayment.
        /// </summary>
        /// <param name="balance">The balance.</param>
        /// <param name="extraRepaymentAmount">The extra repayment amount.</param>
        /// <param name="currentDate">The current date.</param>
        /// <param name="result">The result.</param>
        private void ProcessExtraRepayment(ref float balance, float extraRepaymentAmount, DateTime currentDate, LoanCalculationOutput result)
        {
            // add extra repay transaction (if amount is more than 0)
            balance -= extraRepaymentAmount;
            LoanTransaction transaction = new LoanTransaction(currentDate, TransactionType.ExtraRepayment, extraRepaymentAmount, 0.0f, balance);
            result.Transactions.Add(transaction);
        }

        /// <summary>
        /// Processes the minimum repayment.
        /// </summary>
        /// <param name="balance">The balance.</param>
        /// <param name="context">The context.</param>
        /// <param name="currentDate">The current date.</param>
        /// <param name="monthlyInterest">The monthly interest.</param>
        /// <param name="result">The result.</param>
        private void ProcessMinRepayment(ref float balance, float minRepaymentAmount, DateTime currentDate, float monthlyInterest, LoanCalculationOutput result)
        {
            // add min repay transaction
            if ((balance + monthlyInterest) <= minRepaymentAmount)
            {
                float finalRepayment = 0.0f;
                result.LoanEndsDate = currentDate;
                balance += monthlyInterest;
                finalRepayment = balance;
                balance -= finalRepayment;
                LoanTransaction transaction = new LoanTransaction(currentDate, TransactionType.FinalRepayment, finalRepayment, 0.0f, balance);
                result.Transactions.Add(transaction);
            }
            else
            {
                balance -= minRepaymentAmount;
                LoanTransaction transaction = new LoanTransaction(currentDate, TransactionType.MinimumRepayment, minRepaymentAmount, 0.0f, balance);
                result.Transactions.Add(transaction);
            }
        }

        /// <summary>
        /// Calculates the daily interest.
        /// </summary>
        /// <param name="balance">The balance.</param>
        /// <param name="interestRate">The interest rate.</param>
        /// <returns></returns>
        private float CalculateDailyInterest(float balance, float interestRate)
        {
            float result = 0.0f;
            result = (balance * (interestRate / 100)) / Constants.NUM_DAYS_IN_YEAR;
            return (result);
        }

        /// <summary>
        /// Sets the target end date accuracy.
        /// </summary>
        /// <param name="actualEndDate">The actual end date.</param>
        /// <param name="targetEndDate">The target end date.</param>
        /// <param name="result">The result.</param>
        private void SetTargetEndDateAccuracy(DateTime actualEndDate, DateTime targetEndDate, LoanCalculationOutput result)
        {
            TimeSpan difference = actualEndDate - targetEndDate;
            result.TargetEndDateMissedInDays = Math.Abs((int)difference.TotalDays);
        }

    }
}
